### PR TITLE
Fix false positive for typoed crate or module suggestion

### LINKED ISSUE #76208: Typo'd Imports Result in Confusing Error Message
I wanted to use the chrono crate, so I added it to Cargo.toml and put an import at the top of main.rs: `use chono::prelude::*`, where I typo`d chrono as chono. I got this error message:

```
error[E0433]: failed to resolve: use of undeclared type or module `chono`
 --> src/main.rs:2:5
  |
2 | use chono::prelude::*;
  |     ^^^^^ use of undeclared type or module `chono`
```

I didn't see the typo, and I was confused because I thought I must have put the import in the wrong place, because it said I hadn't declared the module that I was declaring on that line. I checked the docs for E0433, and in the example given someone had forgotten to import the module at all, so I was surprised to see that error on an import line.

### Meta

`rustc --version --verbose`:
```
rustc 1.45.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.45.0
LLVM version: 10.0
```

### LINKED ISSUE #89592: Suggestion for a typoed crate or module results in compile error
This issue is 

I implemented a suggestion for typoed crate or module in #89347. However there are still some edge cases.

Given the following code: 

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=36133bfa20fc000bce25b7b1ea075df9

The current output is:

```
error[E0433]: failed to resolve: use of undeclared crate or module `my_cor`
  --> $DIR/extern-prelude-from-opaque-fail.rs:10:5
   |
LL |     my_cor::mem::drop(0);
   |     ^^^^^^ use of undeclared crate or module `my_cor`
   |
help: there is a crate or module with a similar name
   |
LL |     my_core::mem::drop(0);
   |     ~~~~~~~

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
```
However if you apply this suggestion, you get a compiler error like the following.

```
error[E0433]: failed to resolve: use of undeclared crate or module `my_core`
  --> $DIR/extern-prelude-from-opaque-fail.rs:10:5
   |
LL |     my_core::mem::drop(0);
   |     ^^^^^^^ use of undeclared crate or module `my_core`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
```

Ideally the output should look like:

The compiler shouldn't suggest a crate or module name which you can't use.

```
error[E0433]: failed to resolve: use of undeclared crate or module `my_cor`
  --> $DIR/extern-prelude-from-opaque-fail.rs:10:5
   |
LL |     my_cor::mem::drop(0);
   |     ^^^^^^ use of undeclared crate or module `my_cor`
   |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
```

### PR DESCRIPTION
**Pull request by TaKO8Ki**
follow up of #89347

closes #89592 
closes #76208

r? @estebank

### CODE REVIEW

**Comment by TaKO8Ki** [comment]
I could fix this problem for modules but not for crates. Could you give me some advice?

**Comment by apiraino** [comment]
@TaKO8Ki could you perhaps investigate the CI failing? thanks! :)

@rustbot label -S-waiting-on-review +S-waiting-on-author

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 1]
```diff
@@ -1280,28 +1280,50 @@ impl<'a> Resolver<'a> {
...
+                        *c,
+                        ns,
+                        parent_scope,
+                        false,
+                        c.span,
+                    )
+                    .is_ok()
+                };
+
+                if result { Some(c.name) } else { None }
```
Comment (estebank) [inline compiler/rustc_resolve/src/diagnostics.rs:1321]:
```suggestion
                if *is_extern_crate {
                    Some(c.name)
                } else {
                    let mut ctxt = c.span.ctxt().normalize_to_macros_2_0();
                    let module = self.resolve_self(&mut ctxt, parent_scope.module);
                    self.resolve_ident_in_module(
                        ModuleOrUniformRoot::Module(module),
                        *c,
                        ns,
                        parent_scope,
                        false,
                        c.span,
                    )
                    .ok()
                    .map(|_| c.name)
                }
```

**Review by estebank — COMMENTED** [review]
The changes look reasonable! Apologies for the delay in reviewing. There was a CI failure when I last looked at this PR that needed addressing, and a recent change introduced some merge conflicts. Could you try and rebase on top of master and make sure the tests all succeed?

**Comment by TaKO8Ki** [comment]
@estebank This PR solves a problem for modules but not for extern crates. I tried using `resolve_ident_in_module` for extern crates as well as modules, but it didn't go well. And I have no idea how to solve it. Could you give me some advice?

File: src/test/ui/suggestions/crate-or-module-typo.stderr [hunk 2]
```diff
@@ -1,5 +1,28 @@
+error[E0432]: unresolved import `my_cor`
+  --> $DIR/crate-or-module-typo.rs:31:5
+   |
+LL | use my_cor::mem;
+   |     ^^^^^^ use of undeclared crate or module `my_cor`
+   |
+help: there is a crate or module with a similar name
+   |
+LL | use my_core::mem;
+   |     ~~~~~~~
```
Comment (TaKO8Ki) [inline src/test/ui/suggestions/crate-or-module-typo.stderr:10]: For example, these lines should not be included in stderr.

**Comment by Dylan-DPC** [comment]
@TaKO8Ki any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity

**Comment by estebank** [comment]
@TaKO8Ki feel free to reopen or refile when you have time again!

**Pull request closed without merging**